False positive 'newline-after-var' with 'semi': never

<!--
This template is for bug reports. If you are reporting a bug, please continue on. If you are here for another reason, please see below:

1. To propose a new rule: http://eslint.org/docs/developer-guide/contributing/new-rules
2. To request a change: http://eslint.org/docs/developer-guide/contributing/changes
3. If you have any questions, please stop by our chatroom: https://gitter.im/eslint/eslint

Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->

**What version of ESLint are you using?**
3.2.2

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint

**Please show your full configuration:**
https://github.com/wKich/eslint-config-skbkontur-keforms

**What did you do? Please include the actual source code causing the issue.**
```javascript
const foo = {}

;(foo.bar || []).doSomething()
```

**What did you expect to happen?**
Don't get any warnings or errors.

**What actually happened? Please include the actual, raw output from ESLint.**
Get `newline-after-var` warning

Thanks! I can confirm this as well, with the following config and test file:
```json
{
    "root": true,
    "parserOptions": {
        "ecmaVersion": 6
    },
    "rules": {
        "semi": ["error", "never"],
        "newline-after-var": ["error"]
    }
}

```

```javascript
var foo = bar

;(1 || 2).baz()
```

Results in:
```bash
$ npm run lint

> eslint_bug_tester@1.0.0 lint /Users/vitorbalocco/Workspace/eslint_bug_tester
> eslint src/


/Users/vitorbalocco/Workspace/eslint_bug_tester/src/test.js
  1:1  error  Expected blank line after variable declarations  newline-after-var

✖ 1 problem (1 error, 0 warnings)
```

--- a/lib/rules/newline-after-var.js
+++ b/lib/rules/newline-after-var.js
@@ -46,6 +46,37 @@ module.exports = {
         //--------------------------------------------------------------------------
 
         /**
+         * Gets a token from the given node to compare line to the next statement.
+         *
+         * In general, the token is the last token of the node. However, the token is the second last token if the following conditions satisfy.
+         *
+         * - The last token is semicolon.
+         * - The semicolon is on a different line from the previous token of the semicolon.
+         *
+         * This behavior would address semicolon-less style code. e.g.:
+         *
+         *     var foo = 1
+         *
+         *     ;(a || b).doSomething()
+         *
+         * @param {ASTNode} node - The node to get.
+         * @returns {Token} The token to compare line to the next statement.
+         */
+        function getLastToken(node) {
+            const lastToken = sourceCode.getLastToken(node);
+
+            if (lastToken.type === "Punctuator" && lastToken.value === ";") {
+                const prevToken = sourceCode.getTokenBefore(lastToken);
+
+                if (prevToken.loc.end.line !== lastToken.loc.start.line) {
+                    return prevToken;
+                }
+            }
+
+            return lastToken;
+        }
+
+        /**
          * Determine if provided keyword is a variable declaration
          * @private
          * @param {string} keyword - keyword to test
@@ -114,7 +145,7 @@ module.exports = {
          * @returns {void}
          */
         function checkForBlankLine(node) {
-            const lastToken = sourceCode.getLastToken(node),
+            const lastToken = getLastToken(node),
                 nextToken = sourceCode.getTokenAfter(node),
                 nextLineNum = lastToken.loc.end.line + 1;
 

--- a/tests/lib/rules/newline-after-var.js
+++ b/tests/lib/rules/newline-after-var.js
@@ -224,7 +224,24 @@ ruleTester.run("newline-after-var", rule, {
 
         // should handle one/no blank before case.
         { code: ONE_BLANK_BEFORE_CASE, options: ["always"] },
-        { code: NO_BLANK_BEFORE_CASE, options: ["never"] }
+        { code: NO_BLANK_BEFORE_CASE, options: ["never"] },
+
+        // https://github.com/eslint/eslint/issues/6834
+        {
+            code: `
+                var a = 1
+
+                ;(b || c).doSomething()
+            `,
+            options: ["always"]
+        },
+        {
+            code: `
+                var a = 1
+                ;(b || c).doSomething()
+            `,
+            options: ["never"]
+        },
     ],
 
     invalid: [
@@ -268,6 +285,25 @@ ruleTester.run("newline-after-var", rule, {
         { code: MULTI_LINE_NEXT_LINE_COMMENT, options: ["always"], errors: [ALWAYS_ERROR] },
         { code: MULTI_LINE_NEXT_LINE_BLOCK_COMMENT, options: ["always"], errors: [ALWAYS_ERROR] },
         { code: NEXT_LINE_TWO_COMMENTS_NO_BLANK, options: ["always"], errors: [ALWAYS_ERROR] },
-        { code: NEXT_LINE_COMMENT_BLOCK_COMMENT_NO_BLANK, options: ["always"], errors: [ALWAYS_ERROR] }
+        { code: NEXT_LINE_COMMENT_BLOCK_COMMENT_NO_BLANK, options: ["always"], errors: [ALWAYS_ERROR] },
+
+        // https://github.com/eslint/eslint/issues/6834
+        {
+            code: `
+                var a = 1
+                ;(b || c).doSomething()
+            `,
+            options: ["always"],
+            errors: [ALWAYS_ERROR]
+        },
+        {
+            code: `
+                var a = 1
+
+                ;(b || c).doSomething()
+            `,
+            options: ["never"],
+            errors: [NEVER_ERROR]
+        },
     ]
 });

